doc: GraphQL Generation topic (#58)
ash-project/ash_graphql: documentation/topics/graphql-generation.md

# GraphQL Query Generation

## GraphQL Request and Response

> Following where we left off from [Getting Started with GraphQL](/docs/guides/ash_graphql/latest/tutorials/getting-started-with-graphql), we'll explore what the GraphQL

Reviewer: ```suggestion
Following where we left off from [Getting Started with GraphQL](/documentation/tutorials/getting-started-with-graphql.md), we'll explore what the GraphQL
```

      get :get_ticket, :read 

      # create a field called `list_tickets` that uses the `read` read action to fetch a list of tickets
      list :list_tickets, :read 
    end
  end
```

This time, we've added `list :list_tickets, :read`, to generate a GraphQL query for listing tickets.
The request would look something like this:

```graphql
query {
  list_tickets {
    id
    subject
  }
}
```

And the response would look similar to this:

```json
{
  "data": {
    "list_tickets": [
      {
        "id": "",
        "subject": ""
      }
    ]
  }
}
```

Now, let's say we want to add query parameters to `list_tickets`. How do we do that?
Consider `list :list_tickets, :read` and the `actions` section:

```elixir
  actions do
    # Add a set of simple actions. You'll customize these later.
    defaults [:read, :update, :destroy]
  end

  graphql do
    type :ticket

    queries do
      # create a field called `list_tickets` that uses the `read` read action to fetch a list of tickets
      list :list_tickets, :read 
    end
  end
```

The second argument to `list :list_tickets, :read` is the action that will be called when the query is

> run. In the current example, the action is `:read`, which is the generic Read action. We will need 

Reviewer: ```suggestion
run. In the current example, the action is `:read`, which is the generic Read action. Lets create
a custom action in order to define query parameters for the `list_tickets` query.
```

> run. In the current example, the action is `:read`, which is the generic Read action. We will need 
> to create a custom action in order to define query parameters for the `list_tickets` query.

Reviewer: ```suggestion
run. In the current example, the action is `:read`, which is the generic Read action. Lets create 
a custom action in order to define query parameters for the `list_tickets` query.
```

You can technically take over the primary read, i.e
```elixir
defaults [:update, :destroy]

read :read do
  primary? true
  argument ...
end
```

This just isn't typically a good idea.

We'll call this action `:query_tickets`:

```elixir
  actions do
    defaults [:read, :update, :destroy]
    
    read :query_tickets do
      argument :representative_id, :uuid

      filter(
        expr do
          is_nil(^arg(:representative_id)) or representative_id == ^arg(:representative_id)
        end
      )
    end
  end

  graphql do
    type :ticket

    queries do
      # create a field called `list_tickets` that uses the `:query_tickets` read action to fetch a list of tickets
      list :list_tickets, :query_tickets
    end
  end
```

In the `graphql` section, the `list/2` call has been changed, replacing the `:read` action with `:query_tickets`.

The GraphQL request would look something like this:

```graphql
query($representative_id: ID) {
  list_tickets(representative_id: $representative_id) {
    id
    representative_id
    subject
  }
}
```

## More GraphQL Docs

If you haven't already, please turn on the documentation tag for AshGraphql. Tags can be controlled
at the top of the left navigation menu, under "Including Libraries:".

> - [Getting Started With GraphQL](/docs/guides/ash_graphql/latest/tutorials/getting-started-with-graphql)

Reviewer: So the way that we do links in docs is a bit interesting. We have some tools that transform links, but in this case we'd do it like this
```suggestion
- [Getting Started With GraphQL](/documentation/tutorials/getting-started-with-graphql)
```

> - [AshGraphql API](/docs/dsl/ash_graphql/latest/ashgraphql-api)

Reviewer: ```suggestion
- `AshGraphql.Api`
```
The above will be replaced in hexdocs and ash_hq, although it won't be a link in github (which is the least important of the three targets we have for docs)
